### ISSUE TITLE
The computation complexity of NS iteration and the trade-off of batchsize

### ISSUE THREAD

**Issue by fanghgit** [MoonshotAI/Moonlight#2]
Thanks for sharing your great work! I also have been interested in the optimizer of LLM for quite a well and I am really excited to see people in the industry working on optimizers. I have some questions regarding the computation overhead of the NS iterations.

According to the calculation from [Jordan's blog](https://kellerjordan.github.io/posts/muon/), the overhead of NS iteration is negligible only when the batch size is large enough. For example, if we set batchsize=1, seq_len=32768 and hidden_dim=8192, intermediate size=29568, NS iterations=5, then the overhead will be ~= $$(5 * 6 * 8192 * 29568^2) / ( 1 * 8192 * 32768 * 29568 ) ~= 27.0703125$$. Which is a quite large number.

I guess this mean that we must set the batchsize (or sequence length) to be extremely large in order to amortize the cost of NS iteration. However, extremely large batchsize may hurt the scaling curve. Have you guys considered the trade-off of batchsize?

**Comment by toothacher17** [comment]
Hi, thanks for bringing up this question! Actually we think the scalability for Muon is not a big concern, as we discussed in Distributed Muon. Specifically for your question:

1. bsz = 1 might be a too extreme case. For example, you can see from our paper that we used 4096 examples per global batch. Some other open source work also used very large batch size;

2. because of the ZeRO-1 design, you only need to calculate update for your local stored DP partitioned params, which further reduce the computational needs

The ZeRO-1 based Distributed Muon has a good advantage that as the cluster scales, the master weights are further partitioned and calculation needs are further reduced.

**Comment by fanghgit** [comment]
Thanks for your timely reply, bsz=1 for each DP might be reasonable. Anyway, I agree that ZeRO-1 style can significantly reduce the computation overhead when DP size is large. Thanks for your reply and everything is clear to me now!

**Comment by toothacher17** [comment]
Ah, got u, per DP bsz = 1 possible if you have a lot of DPs. But as you said, when you have a lot of DPs, based on ZeRO-1, you'll still be probably fine.

We'll update the paper with a more detailed analysis soon. Actually this part is the ratio of communication workload is (1, 1.25], and as you have more DPs, you'll get closer to one

**Comment by toothacher17** [comment]
So the actual concern is when you do not have that many DP, but again, when you do not have that many DP, the optimizer consumed time is not significant compared to model's fwd-bwd time

**Comment by JohnHerry** [comment]
请问这个Muon优化器，是主要优化训练速度呢？还是主要优化模型效果？另外除了NLP大模型，在其它模态数据，例如图片和音频数据上也有同样的性能优化效果吗？ 我们一般的要求是训练的时候主要出性能，即模型要能打；推理的时候看速度。优化器一般是训练时候用的，我们可能更关注它是否促进模型性能，对next token prediction提升有多少？
       另外mini-batch大小，如果其他模态的生成模型的话，可能batch大小确实打不到4096那么大。 
       关于ZeRO-1 优化器设计，我纯小白，查了一下deepseek，大概是把梯度分块存储到多个卡上，来节约显存的。但是DDP训练的一个目标，不就是让各个卡上的梯度平均化来促进训练的稳定性吗？如果梯度分步式存储，相当于总计只有一份梯度数据的话，那么DDP的梯度平均化就是放弃了吗？感谢科普一下。

**Comment by fanghgit** [comment]
> 请问这个Muon优化器，是主要优化训练速度呢？还是主要优化模型效果？另外除了NLP大模型，在其它模态数据，例如图片和音频数据上也有同样的性能优化效果吗？ 我们一般的要求是训练的时候主要出性能，即模型要能打；推理的时候看速度。优化器一般是训练时候用的，我们可能更关注它是否促进模型性能，对next token prediction提升有多少？ 另外mini-batch大小，如果其他模态的生成模型的话，可能batch大小确实打不到4096那么大。 关于ZeRO-1 优化器设计，我纯小白，查了一下deepseek，大概是把梯度分块存储到多个卡上，来节约显存的。但是DDP训练的一个目标，不就是让各个卡上的梯度平均化来促进训练的稳定性吗？如果梯度分步式存储，相当于总计只有一份梯度数据的话，那么DDP的梯度平均化就是放弃了吗？感谢科普一下。

我的理解：
- 主要是优化训练速度，优化了训练速度也就意味着在相同的计算量下我们能达到的loss更小，预训练的loss越小也就意味着模型效果越好（预训练时我们不考虑模型泛化性的问题，训练loss越低越好）；
- 对其他模态在原理上也一样，如果Muon能提升与训练的训练速度，那么模型效果也会更好；
- DDP主要是数据并行，模型参数和梯度并没有平均打到每个卡上，比DeepSpeeD ZeRO-1实现下，32卡跑4路DP，每路DP 8卡，也就是模型的参数被分到了8卡上，但是优化器的参数会分到32卡。所以DP数越大优化器的overhead就越小。

**Comment by toothacher17** [comment]
@fanghgit  Thanks for helping answer the questions! I agreed with almost all your answers except **that generalization is not important.**

Actually generalization is very important, otherwise your model might not perform better end to end. One way to derisk is to add some OOD data into your validation data during pretraing. From our experiences, Muon does not have the problem of overfitting too much and hurting the generalization performances.

**Comment by toothacher17** [comment]
@fanghgit 

By the way, we have a more detailed analysis of Distributed Muon's efficiency in the updated pdf version, hope it helps better answer your questions!

**Comment by JohnHerry** [comment]
> > 请问这个Muon优化器，是主要优化训练速度呢？还是主要优化模型效果？另外除了NLP大模型，在其它模态数据，例如图片和音频数据上也有同样的性能优化效果吗？ 我们一般的要求是训练的时候主要出性能，即模型要能打；推理的时候看速度。优化器一般是训练时候用的，我们可能更关注它是否促进模型性能，对next token prediction提升有多少？ 另外mini-batch大小，如果其他模态的生成模型的话，可能batch大小确实打不到4096那么大。 关于ZeRO-1 优化器设计，我纯小白，查了一下deepseek，大概是把梯度分块存储到多个卡上，来节约显存的。但是DDP训练的一个目标，不就是让各个卡上的梯度平均化来促进训练的稳定性吗？如果梯度分步式存储，相当于总计只有一份梯度数据的话，那么DDP的梯度平均化就是放弃了吗？感谢科普一下。
> 
> 我的理解：
> 
> * 主要是优化训练速度，优化了训练速度也就意味着在相同的计算量下我们能达到的loss更小，预训练的loss越小也就意味着模型效果越好（预训练时我们不考虑模型泛化性的问题，训练loss越低越好）；
> * 对其他模态在原理上也一样，如果Muon能提升与训练的训练速度，那么模型效果也会更好；
> * DDP主要是数据并行，模型参数和梯度并没有平均打到每个卡上，比DeepSpeeD ZeRO-1实现下，32卡跑4路DP，每路DP 8卡，也就是模型的参数被分到了8卡上，但是优化器的参数会分到32卡。所以DP数越大优化器的overhead就越小。

感谢！ 您最后一段的意思，是不是说 每个卡上还是有完整的模型参数拷贝， 但是为了提高模型训练“消费数据“的速度，将每8卡上的任务分到同一个DP，同一个DP共享一份完整训练数据，需要对数据分成8块。 但是4个DP之间，是可以做梯度互相参考和平均化的？   还是说，这时候模型的参数也是可以分派到同一个DP的8个卡上的，然后这8个卡的显存支撑起来一份模型参数。整个训练工作，相当于四份模型参数再并行训练？

**Comment by dongdk** [comment]
支持单机单卡吗？

**Comment by toothacher17** [comment]
> 支持单机单卡吗？

@dongdk The toy_train.py in this repo supports running on single GPU

**Comment by dongdk** [comment]
hi @toothacher17 get it. but i try to use it to train a classifier, like clip as backbone, the training speed is slow down. is it normal?
